FinniversKit/Sources/Components/BrazePromotionView/BrazePromotionView.swift

```diff
@@ -0,0 +1,249 @@
+        titleLabel.text = viewModel.title
+        primaryButton.configure(withTitle: viewModel.primaryButtonTitle)
+
+        if let text = viewModel.text {
+            textLabel.text = text
+        } else {
+            textLabel.isHidden = true
+        }
+
+        loadImage()
+        if let dismissible = viewModel.dismissible, !dismissible {
+            closeButton.removeFromSuperview()
+        } else {
+            backgroundView.bringSubviewToFront(closeButton)
+        }
+    }
+
+    private func loadImage() {
+        guard let imageUrl = viewModel.image else {
+            remoteImageView.removeFromSuperview()
+            NSLayoutConstraint.deactivate(stackViewConstraintsImage)
+            NSLayoutConstraint.activate(stackViewConstraintsNoImage)
+            closeButton.setImage(UIImage(named: .cross).withTintColor(.textPrimary), for: .normal)
+            closeButton.backgroundColor = UIColor.clear
+            return
+        }
+        NSLayoutConstraint.deactivate(stackViewConstraintsNoImage)
+        NSLayoutConstraint.activate(stackViewConstraintsImage)
+        backgroundView.addSubview(remoteImageView)
+        remoteImageView.dataSource = imageDatasource
+        NSLayoutConstraint.activate([
+            remoteImageView.widthAnchor.constraint(equalTo: backgroundView.widthAnchor, multiplier: 0.3),
+            remoteImageView.topAnchor.constraint(equalTo: backgroundView.topAnchor),
+            remoteImageView.trailingAnchor.constraint(equalTo: backgroundView.trailingAnchor),
+            remoteImageView.bottomAnchor.constraint(equalTo: backgroundView.bottomAnchor)
+        ])
+        remoteImageView.loadImage(
+            for: imageUrl,
+            imageWidth: 100,
+            fallbackImage: UIImage(named: .noImage)
+        )
+    }
+}
+
+// MARK: - Private functions
+extension BrazePromotionView {
+    private func setup() {
+        translatesAutoresizingMaskIntoConstraints = false
+
+        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(viewWasTapped))
+        addGestureRecognizer(tapGesture)
+
+        addSubview(largeShadowView)
+        largeShadowView.fillInSuperview()
+        largeShadowView.addSubview(smallShadowView)
+        smallShadowView.fillInSuperview()
+        smallShadowView.addSubview(backgroundView)
+        backgroundView.fillInSuperview()
+        backgroundView.addSubview(verticalStackView)
+        backgroundView.addSubview(closeButton)
+        verticalStackView.addArrangedSubviews([titleLabel, textLabel, primaryButton])
+
+        NSLayoutConstraint.activate([
+            closeButton.topAnchor.constraint(equalTo: backgroundView.topAnchor, constant: .spacingS),
+            closeButton.trailingAnchor.constraint(equalTo: backgroundView.trailingAnchor, constant: -.spacingS),
+            closeButton.widthAnchor.constraint(equalToConstant: buttonSize),
+            closeButton.heightAnchor.constraint(equalToConstant: buttonSize)
+        ])
```

Comment:
I personally prefer to group adding of subviews and setting constraints just so it's a bit easier to read. Again, personal preference so it's totally dealer's choice! 🙂

```swift
verticalStackView.addArrangedSubviews([titleLabel, textLabel, primaryButton])

addSubview(largeShadowView)
largeShadowView.addSubview(smallShadowView)
smallShadowView.addSubview(backgroundView)
backgroundView.addSubview(verticalStackView)
backgroundView.addSubview(closeButton)

largeShadowView.fillInSuperview()
smallShadowView.fillInSuperview()
backgroundView.fillInSuperview()

NSLayoutConstraint.activate([
    closeButton.topAnchor.constraint(equalTo: backgroundView.topAnchor, constant: .spacingS),
    closeButton.trailingAnchor.constraint(equalTo: backgroundView.trailingAnchor, constant: -.spacingS),
    closeButton.widthAnchor.constraint(equalToConstant: buttonSize),
    closeButton.heightAnchor.constraint(equalToConstant: buttonSize)
])
```

Reply:
I like to group by view because it feels logical, but I guess your style is like organizing a bookshelf according to color: it looks prettier. My bookshelf at home is organized by color, I will take your suggestion 😊